design/006-renku-api-activities-plans/006-renku-api-activities-plans.md

```diff
@@ -0,0 +1,416 @@
+        +string name
+        +path path
+        +unknown entity
+    }
+    class ProjectStatus{
+        +List-string             stale_outputs
+        +List-renku-api-Activity stale_activities
+        +List-string             modified_inputs
+        +List-string             deleted_inputs
+    }
+    class Plan{
+        +string name
+        +string description
+        +datetime date_created
+        +List-string keywords
+        +List-renku_api_InputField input_fields
+        +List-renku_api_OutputField output_fields
+        +List-renku_api_ParameterField parameter_fields
+        +string command
+        +List-string keywords
+        +List-int success_codes
+        +bool deleted
+        +List-renku_api_Activity activities
+        +List-renku_api_Plan list()
+        
+    }
+    class InputField {
+      +string name
+      +string description
+      +string value
+      +string prefix
+      +int    position
+      +string mapped_stream  
+    }
+    class OutputField {
+      +string name
+      +string description
+      +string value
+      +string prefix
+      +int    position
+      +string mapped_stream  
+    }
+    class ParameterField {
+      +string name
+      +string description
+      +string value
+      +string prefix
+      +int    position
+    }
+    class CompositePlan {
+        +string  name
+        +string  description
+        +datetime  date_created
+        +List-string keywords
+        +List-renku_api_Plan plans
+        +List-renku_api_Mapping mappings
+        +List-renku_api_Link links
+        +bool  deleted
+        +List-renku_api_Activity activities
+    }
+    class  Mapping {
+        <<Connections hidden for readability>>
+        + string name
+        + string description
+        + Path value
+        + renku_api-Input_Output_Parameter_Mappingparameters
+    }
+    class  Link {
+        <<Connections hidden for readability>>
+        +renku_api-Output_Parameter source
+        +List-renku_api-Input_Parameter sinks
+    }
+    class Activity {
+        +datetime started_at
+        +datetime ended_at
+        +string-name-email user
+        +dict annotations
+        +List-renku_api_Input inputs
+        +List-renku_api_Output outputs
+        +List-renku_api_ParameterValue parameters
+        +renku_api_Plan base_plan      
+        +renku_api_Plan executed_plan  
+        +List-renku_api_Activity preceding_activities
+        +List-renku_api_Activity following_activities
+        +List-renku_api_Activity list()
+        +List-renku_api_Activity filter_by_input()
+        +List-renku_api_Activity filter_by_output()
+        +List-renku_api_Activity filter_by_parameter()
+    }
+    class Input {
+      +string checksum
+      +string path
+    }
+    class Output {
+      +string checksum
+      +string path
+    }
+    class ParameterValue {
+      +renku_api-Input_Output_Parameter parameter          
+      +Any value  
+    }
+    Dataset --> DatasetFile : files
```

Comment:
It seems to me that there should be a connection between `Dataset` or `DatasetFile` and `Input`/`Output`. For example, if a user does `Dataset.list()`, they should then be able to get to which activities the dataset was used in. Perhaps, there could be `Dataset.usage -> List[Input, Output]`?

Reply:
We could add that, but we don't have that in the core code at the moment. We could certainly hack it together, like `Dataset.usage = [u for f in dataset.file for a in Activity.filter_by_usage(path=f) for u in Activity.inputs if u.path==f.path]`, though that wouldn't tell you if a file at a commit in the dataset is the same as the file used in an Activity at another commit.

I think this would be more complex and deserves its own RFC, adding the link properly in our internal metadata first plus adding corresponding CLI commands, then thinking about adding it to `renku.api`

Reply:
Yes, that makes sense to me to keep this linking out of scope for this RFC, and leave it to a future RFC 👍